### PR TITLE
tikv: optimize default config for multi-rocksdb

### PR DESCRIPTION
**Pull request by tonyxuqqi**
Signed-off-by: qi.xu <tonxuqi@outlook.com>

### What is changed and how it works?

Issue Number: Ref #12842 

What's Changed:

```commit-message
set block-cache.capacity to 0.3 * memory limit instead of 0.45 for partitioned-raft-kv. 
set region-split-size to 1GB by default when bucket is enabled 
set region-split-size to 10GB by default when partitioned-raft-kv is enabled.

These numbers may be tuned futher if we have other better results.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

```release-note
Default region size now is adjusted according to some features enabled or not.
When region bucket is enabled, the region size is 1GB by default.
When partitioned-raft-kv is enabled, the region size is 10GB by default. 
And of course user can always manually set the region size and it will be respected regardless other settings. 
```

### CODE REVIEW

File: src/storage/config.rs [hunk 1]
```diff
@@ -239,14 +239,18 @@ impl BlockCacheConfig {
         }
     }
 
-    pub fn build_shared_cache(&self) -> Cache {
+    pub fn build_shared_cache(&self, use_multi_rocks: bool) -> Cache {
```
[the remark this answers is not in the document]
Comment (tonyxuqqi) [reply src/storage/config.rs:243]: There's no need to propagate the EngineTypes to everywhere as the concern here is just to know if multi-rocksdb is enabled or not.

[the remark this answers is not in the document]
Comment (tabokie) [reply src/storage/config.rs:243]: engine type is better, most of the places `EngineType` is already imported for other uses. And `tikv_cfg.storage.engine` makes it easy to maintain if we were to update default type in the future.

File: components/raftstore/src/coprocessor/config.rs [hunk 2]
```diff
@@ -100,39 +102,52 @@ impl Default for Config {
 }
 
 impl Config {
+    pub fn region_split_size(&self) -> ReadableSize {
+        self.region_split_size
+            .unwrap_or(if self.enable_region_bucket {
```
Comment (tabokie) [inline components/raftstore/src/coprocessor/config.rs:107]: The preference over `LARGE_REGION_SPLIT_SIZE_MB` or `MULTI_ROCKS_SPLIT_SIZE_MB` is confusing. Needs some comment to say both of these values are used for v1 architecture only.

File: components/raftstore/src/coprocessor/config.rs [hunk 3]
```diff
@@ -100,39 +102,52 @@ impl Default for Config {
...
 
     pub fn region_split_keys(&self) -> u64 {
         // Assume the average size of KVs is 100B.
         self.region_split_keys
-            .unwrap_or((self.region_split_size.as_mb_f64() * 10000.0) as u64)
+            .unwrap_or((self.region_split_size().as_mb_f64() * 10000.0) as u64)
     }
 
-    pub fn validate(&mut self) -> Result<()> {
+    pub fn validate(&mut self, use_multi_rocks: bool) -> Result<()> {
```
Comment (tabokie) [inline components/raftstore/src/coprocessor/config.rs:131]: The old paradigm is to call `optimize_for(type)` first, then `validate`.

File: src/storage/config.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tonyxuqqi) [reply src/storage/config.rs:243]: > engine type is better, most of the places `EngineType` is already imported for other uses. And `tikv_cfg.storage.engine` makes it easy to maintain if we were to update default type in the future.

Fair and good suggestion on "tikv_cfg.storage.engine"

**Review by tabokie — APPROVED** [review]

File: components/raftstore/src/coprocessor/config.rs [hunk 4]
```diff
@@ -100,39 +102,55 @@ impl Default for Config {
...
     }
 
     pub fn region_split_keys(&self) -> u64 {
         // Assume the average size of KVs is 100B.
         self.region_split_keys
-            .unwrap_or((self.region_split_size.as_mb_f64() * 10000.0) as u64)
+            .unwrap_or((self.region_split_size().as_mb_f64() * 10000.0) as u64)
+    }
+
+    pub fn optimize_for(&mut self, use_multi_rocks: bool) {
```
Comment (BusyJay) [inline components/raftstore/src/coprocessor/config.rs:131]:
```suggestion
    pub fn optimize_for(&mut self, raftstore_v2: bool) {
```

**Review by BusyJay — APPROVED** [review]
Rest LGTM

File: src/config/mod.rs [hunk 5]
```diff
@@ -5547,6 +5547,9 @@ mod tests {
             Some(default_cfg.coprocessor.region_split_size() * 3 / 4 / ReadableSize::kb(1));
         default_cfg.raft_store.region_split_check_diff =
             Some(default_cfg.coprocessor.region_split_size() / 16);
+        default_cfg
```
Comment (BusyJay) [inline src/config/mod.rs:5550]: Should optimize first and then query for its size.

Comment (tonyxuqqi) [reply src/config/mod.rs:5550]: In this case it's same result.  But i agree with you.

**Pull request merged**